hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AbstractSqmExpression.java

```diff
@@ -94,7 +94,15 @@ public SqmExpression<String> asString() {
 
 	@Override
 	public <X> SqmExpression<X> as(Class<X> type) {
-		return nodeBuilder().cast( this, type );
+		final SqmExpressible<?> expressible = getExpressible();
+		if ( expressible != null && expressible.getExpressibleJavaType().getJavaTypeClass() == type ) {
+			return (SqmExpression<X>) this;
+		}
+		final BasicType<X> basicTypeForJavaType = nodeBuilder().getTypeConfiguration().getBasicTypeForJavaType( type );
+		if ( basicTypeForJavaType == null ) {
+			throw new IllegalArgumentException( "Can't cast expression to unknown type: " + type.getCanonicalName() );
+		}
+		return new AsWrapperSqmExpression<>( basicTypeForJavaType, this );
```

Comment:
Just like we discussed [here](https://github.com/hibernate/hibernate-orm/pull/8653#issuecomment-2222622818), this should always return a new expression instance.

```suggestion
		final BasicType<X> basicTypeForJavaType = nodeBuilder().getTypeConfiguration().getBasicTypeForJavaType( type );
		if ( basicTypeForJavaType == null ) {
			throw new IllegalArgumentException( "Can't cast expression to unknown type: " + type.getCanonicalName() );
		}
		return new AsWrapperSqmExpression<>( basicTypeForJavaType, this );
```